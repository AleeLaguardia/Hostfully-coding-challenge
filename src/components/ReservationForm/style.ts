import styled from "styled-components";

export const Container = styled.div`
  @keyframes moveIn {
    0% { transform: translateX(-800px) };
    100% { transform: translateX(0) };
  };

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100%;
  width: 100%;
  animation: moveIn 1.3s;
`;

export const Logo = styled.img`
  height: 100px;
  object-fit: cover;
`

export const Content = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;

  .title {
    font-family: ${(props) => props.theme.fonts.poppins};
    font-weight: 500;
    color: ${(props) => props.theme.colors.emperor};
    font-size: ${(props) => props.theme.sizes.m};
  }
`;

export const DateInputContainer = styled.div`
  display: flex;
  width: 100%;
  height: fit-content;
  justify-content: space-between;
  gap: 10px
`;

export const DateInput = styled.div<{ error: boolean }>`
  position: relative;
  display: flex;
  align-items: center;
  height: 40px;
  width: 160px;
  padding: 0 10px;
  background-color: ${(props) => props.theme.colors.white};
  border-radius: 5px;
  border: ${(props) => props.error ? `1px solid ${props.theme.colors.radicalRed}` : 'none'};
  margin: 10px 0;
  cursor: pointer;

  span {
    color: ${(props) => props.theme.colors.purple};
    font-family: ${(props) => props.theme.fonts.poppins};
    font-size: ${(props) => props.theme.sizes.xs};
  }
`;

export const CalendarContainer = styled.div`
  position: absolute;
  display: flex;
  height: fit-content;
  width: fit-content;
  box-shadow: 0px 0px 10px 2px rgba(0,0,0,0.3);
  border-radius: 40px;
  z-index: 3;
  margin-top: 40px;
`;

export const InputContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  width: fit-content;
  height: fit-content;
  margin-top: 50px;
`;

export const ButtonContainer = styled.div`
  display: flex;
  width: 330px;
  height: fit-content;
  justify-content: flex-end;
  margin-top: 20px;
`;