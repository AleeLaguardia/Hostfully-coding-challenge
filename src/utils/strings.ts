export const PATH = {
  HOME: '/',
  BOOKING: '/booking',
  RESERVATION: '/reservation',
};

export const HOME = {
  TITLE_FORM: "Book your stay right now!",
  TITLE_PIC: "Book Your Stay",
  ADULTS: "How many adults",
  CHILDREN: "How many children",
}

export const BOOKING = {
  NO_DATA: "No hotels found in that area",
  ALSO_TRY: "But you can also try:"
};