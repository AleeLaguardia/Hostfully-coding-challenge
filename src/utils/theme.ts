import { DefaultTheme } from "styled-components";

import {
  white,
  alabaster,
  emperor,
  galery,
  gray,
  black,
  purple,
  purpleHover,
  darkPurple,
  radicalRed,
  lightPurple,
  silverChalice,
  alto,
  lynch,
  selectiveYellow,
  chelseaGem,
} from './styling/colors';
import {
  poppins,
  playwrite,
} from './styling/fonts';
import {
  xs,
  xxs,
  xxxs,
  s,
  m,
  l,
  xl,
  xxxl,
} from './styling/sizes';

export const theme: DefaultTheme = {
  colors: {
    white,
    alabaster,
    emperor,
    galery,
    gray,
    black,
    purple,
    purpleHover,
    darkPurple,
    radicalRed,
    lightPurple,
    silverChalice,
    alto,
    lynch,
    selectiveYellow,
    chelseaGem,
  },
  fonts: {
    poppins,
    playwrite,
  },
  sizes: {
    xs,
    xxs,
    xxxs,
    s,
    m,
    l,
    xl,
    xxxl,
  },
}