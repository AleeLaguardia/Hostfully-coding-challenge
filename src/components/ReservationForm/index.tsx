import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { setUser } from "../../store/slice/userInfoSlice";
import LogoIcon from '../../assets/purple-logo-1.png'
import Button from "../Button";
import Dropdown from "../Dropdown";
import { ButtonContainer, CalendarContainer, Container, Content, DateInput, DateInputContainer, InputContainer, Logo } from "./style";
import { HOME, PATH } from "../../utils/strings";
import CalendarComponent from "../CalendarComponent";
import { RootState } from "../../store";
import { differenceInDays, format, setDate } from "date-fns";
import useClickOutside from "../../utils/hooks/useClickOutside";
import { useNavigate } from "react-router-dom";
import Input from "../Input";
import { theme } from "../../utils/theme";

const optionsAdults = ['1 adult', '2 adults', '3 adults', '4 adults', '5 adults', 'More than 5'];
const optionsChildren = ['No children', '1 child', '2 children', '3 children', '4 children', '5 children', 'More than 5'];

const ReservationForm: React.FC = () => {
  const [destination, setDestination] = useState<string>('');
  const [adults, setAdults] = useState<string>('');
  const [children, setChildren] = useState<string>('');

  const [isDropdownAdultOpen, setIsDropdownAdultOpen] = useState<boolean>(false);
  const [isDropdownChildrenOpen, setIsDropdownChildrenOpen] = useState<boolean>(false);
  const [isCalendarOpen, setIsCalendarOpen] = useState<boolean>(false);

  const [destinationError, setDestinationError] = useState<boolean>(false);
  const [dateError, setDateError] = useState<boolean>(false);
  const [adultsError, setAdultsError] = useState<boolean>(false);
  const [childrenError, setChildrenError] = useState<boolean>(false);

  const dispatch = useDispatch();
  const { date } = useSelector((state: RootState) => state.user);

  const navigate = useNavigate();

  const calendarRef = useClickOutside(() => setIsCalendarOpen(false));
  const adultDropdownRef = useClickOutside(() => setIsDropdownAdultOpen(false));
  const childrenDropdownRef = useClickOutside(() => setIsDropdownChildrenOpen(false));

  const handleInputChange = (
    e: React.ChangeEvent<HTMLInputElement>,
    setState: React.Dispatch<React.SetStateAction<string>>
  ) => {
    e.preventDefault();
    setState(e.target.value);
  };

  const handleConfirm = () => {
    if (destination === '') {
      setDestinationError(true);
    } else {
      setDestinationError(false);
    }

    if (adults === '') {
      setAdultsError(true);
    } else {
      setAdultsError(false);
    }

    if (children === '') {
      setChildrenError(true)
    } else {
      setChildrenError(false)
    }

    if (differenceInDays(date[1], date[0]) < 1) {
      setDateError(true);
    } else {
      setDateError(false);
    }

    if (destination.length !== 0 && adults.length !== 0 && children.length !== 0 && differenceInDays(date[1], date[0]) > 0) {
      dispatch(setUser({ destination, date, adults, children }));
      navigate(`${PATH.BOOKING}?destination=${destination}&checkin=${format(date[0], 'MM/dd/yyyy')}&checkout=${format(date[1], 'MM/dd/yyyy')}&adults=${adults}&children=${children}`);
    }
  }

  return (
    <Container>
      <Content>
        <Logo draggable={false} src={LogoIcon} alt="LogoIcon" />
        <InputContainer>
          <span className="title">{HOME.TITLE_FORM}</span>
          <Input
            placeholder="Destination..."
            value={destination}
            onChange={(e) => handleInputChange(e, setDestination)}
            error={destinationError}
          />
          <DateInputContainer ref={calendarRef}>
            <DateInput data-testid="first-date-calendar-dataid" onClick={() => setIsCalendarOpen(true)} error={dateError}>
              <span>{format(date[0], 'MM/dd/yyyy')}</span>
            </DateInput>
            <DateInput data-testid="second-date-calendar-dataid" onClick={() => setIsCalendarOpen(true)} error={dateError}>
              <span>{format(date[1], 'MM/dd/yyyy')}</span>
            </DateInput>
            {isCalendarOpen && (
              <CalendarContainer>
                <CalendarComponent isOpen={isCalendarOpen} setIsOpen={setIsCalendarOpen} />
              </CalendarContainer>
            )}
          </DateInputContainer>
          <DateInputContainer ref={isDropdownAdultOpen ? adultDropdownRef : childrenDropdownRef}>
            <Dropdown
              error={adultsError}
              data-testid="adult-dropdown"
              placeholder={HOME.ADULTS}
              isDropdownOpen={isDropdownAdultOpen}
              setIsDropdownOpen={setIsDropdownAdultOpen}
              setOptionSelected={setAdults}
              optionSelected={adults}
              options={optionsAdults}
            />
            <Dropdown
              error={childrenError}
              data-testid="children-dropdown"
              placeholder={HOME.CHILDREN}
              isDropdownOpen={isDropdownChildrenOpen}
              setIsDropdownOpen={setIsDropdownChildrenOpen}
              setOptionSelected={setChildren}
              optionSelected={children}
              options={optionsChildren}
            />
          </DateInputContainer>
        </InputContainer>
        <ButtonContainer>
          <Button
            onClick={handleConfirm}
            label="Confirm"
          />
        </ButtonContainer>
      </Content>
    </Container>
  );
};

export default ReservationForm;
