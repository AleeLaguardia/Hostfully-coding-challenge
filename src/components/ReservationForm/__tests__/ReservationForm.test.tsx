import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import ReservationForm from '..';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../../utils/theme';
import { BrowserRouter } from 'react-router-dom';
import { differenceInDays } from 'date-fns';

const mockStore = configureStore([]);

describe('ReservationForm', () => {
  let store: any;

  beforeEach(() => {
    store = mockStore({
      user: {
        destination: '',
        adults: '',
        children: '',
        date: [new Date('2024-07-14'), new Date('2024-07-17')],
      },
    });
  });

  it('renders ReservationForm correctly', () => {
    const { getByText, getByPlaceholderText, asFragment } = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <ReservationForm />
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    );

    expect(getByPlaceholderText('Destination...')).toBeInTheDocument();
    expect(getByText('How many adults')).toBeInTheDocument();
    expect(getByText('How many children')).toBeInTheDocument();
    expect(getByText('Confirm')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles input change correctly', () => {
    const { getByPlaceholderText, getByText, getByTestId } = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <ReservationForm />
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    );
    
    fireEvent.click(getByTestId('first-date-calendar-dataid'));

    fireEvent.click(getByText('16'));
    fireEvent.click(getByText('18'));

    fireEvent.click(getByText('Confirm'));

    expect(getByPlaceholderText('Destination...')).toHaveStyle('border: 1px solid #FA1E71');

    fireEvent.change(getByPlaceholderText('Destination...'), { target: { value: 'Paris' } });

    expect(getByPlaceholderText('Destination...')).toHaveValue('Paris');
  });

  it('handles confirmation correctly', async () => {
    const { getByText, getByPlaceholderText, getByTestId } = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <ReservationForm />
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    );

    fireEvent.change(getByPlaceholderText('Destination...'), { target: { value: 'Paris' } });
    fireEvent.click(getByTestId('first-date-calendar-dataid'));
    fireEvent.click(getByText('14'));
    fireEvent.click(getByTestId('second-date-calendar-dataid'));
    fireEvent.click(getByText('17'));

    fireEvent.click(getByTestId('adult-dropdown'));
    fireEvent.click(getByText('2 adults'));

    fireEvent.click(getByTestId('children-dropdown'));
    fireEvent.click(getByText('1 child'));

    fireEvent.click(getByText('Confirm'));

    await waitFor(() => {
      const actions = store.getActions();
      expect(actions).toEqual([
        {
          type: 'user/setUser',
          payload: {
            destination: 'Paris',
            date: [new Date('2024-07-14T00:00:00.000Z'), new Date('2024-07-17T00:00:00.000Z')],
            adults: '2 adults',
            children: '1 child',
          },
        },
      ]);
    });
  });

  it('handles dropdown selections correctly', () => {
    const { getByTestId, getByText } = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <ReservationForm />
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    );

    fireEvent.click(getByTestId('adult-dropdown'));
    fireEvent.click(getByText('2 adults'));

    expect(getByTestId('adult-dropdown')).toHaveTextContent('2 adults');

    fireEvent.click(getByTestId('children-dropdown'));
    fireEvent.click(getByText('1 child'));

    expect(getByTestId('children-dropdown')).toHaveTextContent('1 child');
  });
});
