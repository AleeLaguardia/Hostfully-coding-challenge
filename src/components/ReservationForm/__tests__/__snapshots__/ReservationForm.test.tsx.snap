// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReservationForm renders ReservationForm correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-iGgVNO ctVDoY"
  >
    <div
      class="sc-kAycRU gTzSLr"
    >
      <img
        alt="LogoIcon"
        class="sc-gsFSjX ckeHlI"
        draggable="false"
        src="purple-logo-1.png"
      />
      <div
        class="sc-kpDprT lhLlMS"
      >
        <span
          class="title"
        >
          Book your stay right now!
        </span>
        <input
          class="sc-cwHqhk cBSfcL"
          placeholder="Destination..."
          value=""
        />
        <div
          class="sc-imWZod bqizog"
        >
          <div
            class="sc-jXbVAB bQtoyp"
            data-testid="first-date-calendar-dataid"
          >
            <span>
              07/13/2024
            </span>
          </div>
          <div
            class="sc-jXbVAB bQtoyp"
            data-testid="second-date-calendar-dataid"
          >
            <span>
              07/16/2024
            </span>
          </div>
        </div>
        <div
          class="sc-imWZod bqizog"
        >
          <div
            class="sc-gEvDqW jGvbiZ"
            data-testid="adult-dropdown"
          >
            <div
              class="sc-eqUzNf iGyBLM"
            >
              <span>
                How many adults
              </span>
              <img
                alt="ArrowIcon"
                class="sc-fqkwJk hXoYFT"
                src="arrow.svg"
              />
            </div>
          </div>
          <div
            class="sc-gEvDqW jGvbiZ"
            data-testid="children-dropdown"
          >
            <div
              class="sc-eqUzNf iGyBLM"
            >
              <span>
                How many children
              </span>
              <img
                alt="ArrowIcon"
                class="sc-fqkwJk hXoYFT"
                src="arrow.svg"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-dAlxHm enLAMC"
      >
        <button
          class="sc-aYaIB jbXBCB"
          color="#5F259E"
          colorhover="#451B74"
          labelcolor="#FFFFFF"
          width=""
        >
          <span>
            Confirm
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
